src/mesh/surface_interpolant.jl

```diff
@@ -1,53 +1,45 @@
-  x                   ::SVector{NNODES, RealT}
-  y                   ::SVector{NNODES, RealT}
-end
-
-
-# construct a single instance of the gamma curve struct
-function GammaCurve(RealT, curve_polydeg, curve_x_vals, curve_y_vals)
-  nnodes_ = curve_polydeg + 1
-
-  nodes, _ = chebyshev_gauss_lobatto_nodes_weights(nnodes_)
-  wbary    = barycentric_weights(nodes)
-
-  return GammaCurve{RealT, nnodes_}(nodes, wbary, curve_x_vals, curve_y_vals)
+struct CurvedSurface{RealT<:Real, NNODES}
+  nodes               ::Vector{RealT}
+  barycentric_weights ::Vector{RealT}
+  coordinates         ::Array{RealT, 2}
 end
 
 
 # evalute the Gamma curve interpolant at a particular point s and return the (x,y) coordinate
-function evaluate_at(s, boundary_curve::GammaCurve)
+function evaluate_at(s, boundary_curve::CurvedSurface)
 
-   @unpack nodes, barycentric_weights, x, y = boundary_curve
+   @unpack nodes, barycentric_weights, coordinates = boundary_curve
 
-   x_coordinate_at_s_on_boundary_curve = lagrange_interpolation(s, nodes, x, barycentric_weights)
-   y_coordinate_at_s_on_boundary_curve = lagrange_interpolation(s, nodes, y, barycentric_weights)
+   x_coordinate_at_s_on_boundary_curve = lagrange_interpolation(s, nodes, view(coordinates, 1, :),
+                                                                barycentric_weights)
+   y_coordinate_at_s_on_boundary_curve = lagrange_interpolation(s, nodes, view(coordinates, 2, :),
+                                                                barycentric_weights)
 
    return x_coordinate_at_s_on_boundary_curve, y_coordinate_at_s_on_boundary_curve
 end
 
 
 # evalute the derivative of a Gamma curve interpolant at a particular point s
 # and return the (x,y) coordinate
-function derivative_at(s, boundary_curve::GammaCurve)
+function derivative_at(s, boundary_curve::CurvedSurface)
 
-   @unpack nodes, barycentric_weights, x, y = boundary_curve
+   @unpack nodes, barycentric_weights, coordinates = boundary_curve
 
-   x_coordinate_at_s_on_boundary_curve_prime = lagrange_interpolation_derivative(s, nodes, x,
-                                                                                 barycentric_weights)
-   y_coordinate_at_s_on_boundary_curve_prime = lagrange_interpolation_derivative(s, nodes, y,
-                                                                                 barycentric_weights)
+   x_coordinate_at_s_on_boundary_curve_prime = lagrange_interpolation_derivative(s, nodes,
+                                                                             view(coordinates, 1, :),
+                                                                             barycentric_weights)
+   y_coordinate_at_s_on_boundary_curve_prime = lagrange_interpolation_derivative(s, nodes,
+                                                                             view(coordinates, 2, :),
+                                                                             barycentric_weights)
```

Comment:
I am open to either one. I chose `ndims, nnodes` for consistency with the other containers in `solvers`. But since these two features are disconnected we can pick whichever one would be more performant

Reply:
You could run the benchmark code I posted above and see whether swapping the order makes a difference here. Then, you could pick the ore efficient version and resolve this issue.

Reply:
I attempted to do benchmarking of `evaluate_at` but it kept throwing an error that `isapprox` was unknown

> julia> using Revise, BenchmarkTools; using Trixi
> 
> julia> trixi_include("examples/2d/elixir_euler_unstructured_quad_periodic.jl")
> 
> julia> @ benchmark Trixi.evaluate_at($(Float64), $(mesh.surface_curves[2,1]))
> ERROR: MethodError: no method matching isapprox(::Type{Float64}, ::Float64; rtol=2.220446049250313e-16)
> Closest candidates are:
>   isapprox(::Missing, ::Any; kwargs...) at missing.jl:90
>   isapprox(::Number, ::Number; atol, rtol, nans, norm) at floatfuncs.jl:278
>   isapprox(::Any; kwargs...) at floatfuncs.jl:291
>   ...
> Stacktrace:
>   [1] lagrange_interpolation(x::Type, nodes::Vector{Float64}, fvals::SubArray{Float64, 1, Matrix{Float64}, Tuple{Int64, Base.Slice{Base.OneTo{Int64}}}, true}, wbary::Vector{Float64})
>     @ Trixi ~/Documents/source_code/julia/Trixi.jl/src/mesh/surface_interpolant.jl:76
>   [2] evaluate_at(s::Type, boundary_curve::Trixi.CurvedSurface{Float64})
>     @ Trixi ~/Documents/source_code/julia/Trixi.jl/src/mesh/surface_interpolant.jl:22
>   [3] var"##core#453"(Float64#451::Type{Float64}, 452::Trixi.CurvedSurface{Float64})
>     @ Main ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:479
>   [4] var"##sample#454"(__params::BenchmarkTools.Parameters)
>     @ Main ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:485
>   [5] _run(b::BenchmarkTools.Benchmark, p::BenchmarkTools.Parameters; verbose::Bool, pad::String, kwargs::Base.Iterators.Pairs{Symbol, Integer, NTuple{4, Symbol}, NamedTuple{(:samples, :evals, :gctrial, :gcsample), Tuple{Int64, Int64, Bool, Bool}}})
>     @ BenchmarkTools ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:98
>   [6] #invokelatest#2
>     @ ./essentials.jl:710 [inlined]
>   [7] #run_result#38
>     @ ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:33 [inlined]
>   [8] run(b::BenchmarkTools.Benchmark, p::BenchmarkTools.Parameters; progressid::Nothing, nleaves::Float64, ndone::Float64, kwargs::Base.Iterators.Pairs{Symbol, Integer, NTuple{5, Symbol}, NamedTuple{(:verbose, :samples, :evals, :gctrial, :gcsample), Tuple{Bool, Int64, Int64, Bool, Bool}}})
>     @ BenchmarkTools ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:116
>   [9] #warmup#47
>     @ ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:168 [inlined]
>  [10] warmup(item::BenchmarkTools.Benchmark)
>     @ BenchmarkTools ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:168
>  [11] top-level scope
>     @ ~/.julia/packages/BenchmarkTools/ms0Xc/src/execution.jl:387

Reply:
> julia> @ benchmark Trixi.evaluate_at($(Float64), $(mesh.surface_curves[2,1]))
With `$(Float64)` you do not insert a value `s`, but the type `Float64`. Maybe try
```julia
julia> @ benchmark Trixi.evaluate_at($(0.0), $(mesh.surface_curves[2,1]))
```

Reply:
Your last command should be something like
```julia
julia> @benchmark Trixi.evaluate_at(xi, $(mesh.surface_curves[2,1])) setup=(xi=2 * rand() - 1)
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     253.175 ns (0.00% GC)
  median time:      253.616 ns (0.00% GC)
  mean time:        256.440 ns (0.00% GC)
  maximum time:     535.744 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     359
```
You need to pass a value of type `Flaot64`, not the type itself. You want to evaluate at, say, `0.5`, not at `Float64`.

Reply:
Okay, as @ranocha predicted there is a gain by swapping the order. This used `mesh_periodic_square_with_twist.mesh`.

ordering with `[ndims, nnodes]`:
```julia
julia> @benchmark Trixi.evaluate_at($(0.1), $(mesh.surface_curves[2,1]))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     436.737 ns (0.00% GC)
  median time:      437.338 ns (0.00% GC)
  mean time:        459.978 ns (0.00% GC)
  maximum time:     2.338 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     198

 julia> @benchmark Trixi.derivative_at($(0.0), $(mesh.surface_curves[2,1]))
  BenchmarkTools.Trial: 
    memory estimate:  0 bytes
    allocs estimate:  0
    --------------
    minimum time:     304.780 ns (0.00% GC)
    median time:      308.044 ns (0.00% GC)
    mean time:        313.582 ns (0.00% GC)
    maximum time:     1.398 μs (0.00% GC)
    --------------
    samples:          10000
    evals/sample:     250
```
   
ordering with `[nnodes, ndims]`:
```julia
julia> @benchmark Trixi.evaluate_at($(0.1), $(mesh.surface_curves[2,1]))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     463.675 ns (0.00% GC)
  median time:      464.036 ns (0.00% GC)
  mean time:        471.551 ns (0.00% GC)
  maximum time:     1.233 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     197

julia> @benchmark Trixi.derivative_at($(0.0), $(mesh.surface_curves[2,1]))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     325.697 ns (0.00% GC)
  median time:      329.167 ns (0.00% GC)
  mean time:        342.774 ns (0.00% GC)
  maximum time:     1.322 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     228
```